File: Java/src/Medium/q15.java
package Medium;

import java.util.*;

public class q15 {
    public static List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> temp;

        int left, right;
        for (int i=0;i<nums.length;++i) {
            if (nums[0] > 0)
                return result;

            if (i > 0 && nums[i] == nums[i-1])
                continue;

            left = i+1;
            right = nums.length-1;

            while (left < right) {
                if (nums[i]+nums[left]+nums[right] > 0)
                    right--;
                else if (nums[i]+nums[left]+nums[right] < 0)
                    left++;
                else {
                    temp = new ArrayList<>();
                    temp.add(nums[i]);
                    temp.add(nums[left]);
                    temp.add(nums[right]);

                    result.add(temp);

                    // Remove duplicated b,c
                    while (left < right && nums[left] == nums[left+1])
                        left++;

                    while (left < right && nums[right] == nums[right-1])
                        right--;

                    right--;
                    left++;
                }
            }
        }

        return result;
    }

    public static void testSample() {
        List<List<Integer>> result = threeSum(new int[]{-1,0,1,2,-1,-4});

        System.out.println(result);
    }
}

File: Python/src/medium/q15.py
from typing import List

class Solution:
    """
    Leetcode 15: 3Sum
    """

    def threeSum(self, nums: list[int]) -> list[list[int]]:
        nums.sort() # O(nlogn)
        result = []

        for i in range(len(nums)): # O(n)
            if i > 0 and nums[i] == nums[i-1]:
                continue

            j, k = i+1, len(nums)-1

            while j < k: # O(n)
                t_sum = nums[i] + nums[j] + nums[k]

                if t_sum > 0:
                    k -= 1
                elif t_sum < 0:
                    j += 1
                else:
                    result.append([nums[i], nums[j], nums[k]])
                    j += 1
                    while nums[j] == nums[j-1] and j < k:
                        j += 1
        
        return result
    

if __name__ == "__main__":
    sol = Solution()
    print(sol.threeSum([0,0,0,0]))